More unified language support

satyamtg opened the pull request:
Fixes #18 by using iso-639 to basically return language details if available for any type of input. For example, the following inputs would give details for Hindi language -
- hi
- hin
- Hindi

The returned details are in the form of a dictionary similar to the one below - 
```
{
                "iso-639-1": "zh",
                "iso-639-2b": "chi",
                "iso-639-2t": "zho",
                "iso-639-3": "zho",
                "iso-639-5": "",
                "english": "Chinese",
                "iso_types": ["part1"],
                "querytype": "locale",
                "query": "zh-Hans",
}
```

The iso_types holds a list of ISO639 collections to which the query part matches

This breaks the current API

satyamtg commented on src/zimscraperlib/i18n.py:
```diff
@@ -48,51 +50,90 @@ def setlocale(root_dir, locale_name):
-        )
-    except iso639.NonExistentLanguageError:
+def get_language_details(query):
+    """returns language details if query is a language or a locale otherwise returns None"""
```
Resolved

satyamtg commented:
> we should look for the `macro` of a language and if it's there (it's a sub lang of a macro, like most chineses and arabics) fill the other info like 639-1 and such via the macro.

Have implemented this. This should now work fine.



> Next, I don't understand why we lost the native language name. I think it's a very important information. We should add another dep. to fetch it if iso-639 doesn't provide it.
> It's too bad because it bundles the info in a `languages_utf-8.txt` file it doesn't use…

I did this due to basically 2 things -
1. The iso-639 library doesn't have native languages. The other one that we previously used i.e. iso639 has them in a languages_utf-8.txt file and that too doesn't have that for all languages listed there. It only has support for iso-639-2 languages.

2. Both the dependencies have the same package name. So, they are not compatible with each other and pip would rewrite the one previously installed. Also, iso639 does everything in \_\_init\_\_.py and that's not good.

So, IMO, what we should do is to implement our own list of native languages in this scraperlib as its important info (and we would be able to add other languages too. What do you say @rgaudin ?




> Do we want to use `und` when we can't parse the source? I'm all for explicit fails it might make sense in some contexts (via a `failsafe` arg?)

It returns None if we don't get a result. IMO it's good to fail if we cannot parse the source query. It may be useful with a failsafe arg though. Will add this in next commit.

rgaudin commented:
> So, IMO, what we should do is to implement our own list of native languages in this scraperlib as its important info (and we would be able to add other languages too. What do you say @rgaudin ?

I think we can find this elsewhere. I'd rather we don't have to do it ourselves. `pycountry` came to my mind but doesn't does this. We can search for another dep.
I stumbled upon https://github.com/umpirsky/language-list which have some interesting info but it's a slightly different usage.
 
> > Do we want to use `und` when we can't parse the source? I'm all for explicit fails it might make sense in some contexts (via a `failsafe` arg?)
> 
> It returns None if we don't get a result. IMO it's good to fail if we cannot parse the source query. It may be useful with a failsafe arg though. Will add this in next commit.

OK, agrees.

satyamtg commented:
> I stumbled upon https://github.com/umpirsky/language-list which have some interesting info but it's a slightly different usage.

I see that this has a list of 609 languages (with both 2-lettered and 3-lettered codes) in 564 locales. However, all 609 languages are not named in all 564 locales. I also came across [this](https://github.com/mozilla/language-mapping-list/blob/master/language-mapping-list.js) from Mozilla that has over 200 languages named in native script (with locales) which would be definitely easier to extract as that's a valid JSON which is returned.

rgaudin commented:
I think we should go with `babel`. It seems quite complete.

It works with locale names and ISO codes but not all of them. For example, the codes for the subs of arabic macro don't work… but the country locales do (`arq` fails but `ar_MA` works).
For missing values, if it has a macro, we'd get the one of the macro and if it's not, we'd use the English string.

```py
In [2]: from babel import Locale

In [3]: Locale.parse("zh_TW").get_display_name()
Out[3]: '中文 (繁體, 台灣)'

In [4]: Locale.parse("fr").get_display_name()
Out[4]: 'français'

In [5]: Locale.parse("fra").get_display_name()
Out[5]: 'français (France)'

In [6]: Locale.parse("bam").get_display_name()
Out[6]: 'bamanakan (Mali)'

In [7]: Locale.parse("bm").get_display_name()
Out[7]: 'bamanakan'

In [8]: Locale.parse("ara").get_display_name()
Out[8]: 'العربية (مصر)'

In [9]: Locale.parse("ar").get_display_name()
Out[9]: 'العربية'

In [10]: Locale.parse("cmn").get_display_name()
Out[10]: '中文 (简体, 中国)'

In [11]: Locale.parse("arq").get_display_name()
---------------------------------------------------------------------------
UnknownLocaleError                        Traceback (most recent call last)
<ipython-input-11-2464e6e4c637> in <module>
----> 1 Locale.parse("arq").get_display_name()

~/src/envs/zimscraperlib/lib/python3.8/site-packages/babel/core.py in parse(cls, identifier, sep, resolve_likely_subtags)
    329                 return locale
    330
--> 331         raise UnknownLocaleError(input_id)
    332
    333     def __eq__(self, other):

UnknownLocaleError: unknown locale 'arq'

In [12]: Locale.parse("arb").get_display_name()
Out[12]: 'العربية (مصر)'

In [13]: Locale.parse("ary").get_display_name()
---------------------------------------------------------------------------
UnknownLocaleError                        Traceback (most recent call last)
<ipython-input-13-de5bd4e24796> in <module>
----> 1 Locale.parse("ary").get_display_name()

~/src/envs/zimscraperlib/lib/python3.8/site-packages/babel/core.py in parse(cls, identifier, sep, resolve_likely_subtags)
    329                 return locale
    330
--> 331         raise UnknownLocaleError(input_id)
    332
    333     def __eq__(self, other):

UnknownLocaleError: unknown locale 'ary'

In [14]: Locale.parse("ar-MA").get_display_name()
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-14-39398b209e9b> in <module>
----> 1 Locale.parse("ar-MA").get_display_name()

~/src/envs/zimscraperlib/lib/python3.8/site-packages/babel/core.py in parse(cls, identifier, sep, resolve_likely_subtags)
    266             raise TypeError('Unexpected value for identifier: %r' % (identifier,))
    267
--> 268         parts = parse_locale(identifier, sep=sep)
    269         input_id = get_locale_identifier(parts)
    270

~/src/envs/zimscraperlib/lib/python3.8/site-packages/babel/core.py in parse_locale(identifier, sep)
   1092     lang = parts.pop(0).lower()
   1093     if not lang.isalpha():
-> 1094         raise ValueError('expected only letters, got %r' % lang)
   1095
   1096     script = territory = variant = None

ValueError: expected only letters, got 'ar-ma'

In [15]: Locale.parse("ar_MA").get_display_name()
Out[15]: 'العربية (المغرب)'
```

satyamtg commented:
> I think we should go with [`babel`](http://babel.pocoo.org). It seems quite complete.
> 
> It works with locale names and ISO codes but not all of them. For example, the codes for the subs of arabic macro don't work… but the country locales do (`arq` fails but `ar_MA` works).
> For missing values, if it has a macro, we'd get the one of the macro and if it's not, we'd use the English string.

Yup. Babel is in-fact the best solution we can have for this. Thanks for pointing that out. Implemented the native language part from it and added tests.
